docs/design/README.md

```diff
@@ -0,0 +1,595 @@
+The command creates a directory structure like the one shown above.
+The new project will include a minimal variant with a simple custom package and settings.
+The command will take a few arguments, such as `--project name` and `--variant-name`.
+
+Similarly, a command will be available to add a new variant to an existing project.
+For example, `twoliter new-variant` which will add a variant.
+
+```
+❯ twoliter new project --name ootb-variants --variant-name example-dev
+❯ cd ootb-variants
+❯ twoliter build variant --name example-dev --arch x86_64
+  ...
+❯ twoliter new variant --name my-new-variant
+```
+
+### Twoliter Update
+
+**TODO** - This section is a bit hand-wavy and will have its design further refined.
+See https://github.com/bottlerocket-os/twoliter/issues/4
+
+`twoliter update` is another inspiration from `cargo`, this time inspired by `cargo update`.
+This command makes it easy for maintainers to update the packages (kits actually) they are using to the latest versions.
+
+In the original Bottlerocket build system, we used Cargo crates to model RPM dependencies.
+In Twoliter, we will again lean on Cargo to manage dependencies for us.
+We will create a local Cargo registry to represent kit and SDK versions, then use `cargo update` to resolve the latest dependencies.
+
+Before describing this procedure, we should consider that no two versions of the same kit can be in the dependency tree.
+Since they are Yum repositories, this would mess things up.
+Cargo allows multiple major versions of a package to exist in a build, but this is behavior we do not want.
+We will manipulate version numbers to so that Cargo gives us the desired behavior.
+
+First Twoliter will read the kit dependencies from each variant and add any external kits to its graph.
+Next Twoliter will read the kit definitions in the `kits` directory, and add all of their external dependencies to its graph.
+
+Once it has this list, Twoliter will begin a traversal.
+For each kit it will pull all the `-metadata` images for each external kit JSON files locally.
+Each time it finds a new external kit (i.e. an external kit that is depended on by an external kit), it will add this new dependency to the graph.
+
+Once it has all the dependency and version information, Twoliter will build a local cargo registry.
+Each kit will be represented by an empty Cargo `lib.rs` project with the same name as the kit.
+The Cargo package version will be given by deterministically adding the kit's version to 1.x (**TODO** - hand-waved).
+
+The SDK will be a special Cargo package.
+Two SDKs may have the same version, but come from different container registries.
+To ensure a single SDK is used, it will be given a version in Cargo like this (**TODO** - or some hash of the necessary unique information):
+
+```toml
+[dependencies]
+sdk = "=0.50.0-public.ecr.aws_bottlerocket_bottlerocket-sdk-x86_64"
```

Comment:
I definitely don't want "arch" to appear here. I'd like that to be something the system resolves automatically to reduce the friction for adding a new architecture - going from `x86_64` only to also supporting `aarch64`, for example.

I'd also suggest having some sort of default here where everything is assumed to be rooted under `public.ecr.aws/bottlerocket` unless otherwise specified. So:

```
[dependencies]
# implicitly gets modified to append the target arch during dependency resolution
# results in public.ecr.aws/bottlerocket/bottlerocket-sdk-x86_64:v0.50.0
bottlerocket-sdk = "0.50.0"

# explicitly set to some other prefix
my-kit = { source = "my.registry/prefix", version = "0.10.0" }
```